### ISSUE TITLE
Create utility to properly traverse validators/reward accounts + tests

### ISSUE THREAD

**Issue by fernandomg** [poanetwork/poa-bridge-contracts#123]
To properly handle validators and reward addresses, to distribute `fee` among them, a double linked list will be helpful.

See https://github.com/poanetwork/poa-bridge-contracts/issues/119#issuecomment-448246916 for further details.

**Comment by akolotov** [comment]
> we couldn't find a suitable use for the Prev list, as we can add/remove elements from one single linked list without issues.

Actually we could measure how many gas two SSTORE operations to update the `Prev` list will consume. And compare it with the `while`-loop you are using. I think for ten or even twenty validators the loop still makes sense. But you can confirm this by measurements. 
The idea was to not look the parent element of the needed validator through the list but get it by using the `Prev` list.

```
function removeValidatorFromLinkedList(address _validator) internal {
    address memory validatorNext = Next[_validator];
    address memory validatorPrev = Prev[_validator];

    Next[validatorPrev]=validatorNext;
    Prev[validatorNext]=validatorPrev;
    delete Next[_validator];
    delete Prev[_validator];
}
```

**Comment by akolotov** [comment]
Probably, it is good idea to use an address 0xffffffffffffffffffffffffffffffffffffffff as the marker of the end of the list (as well as the index of the first element) instead of `address(0)` since `Next[<non-existing element> == address(0)` returns `true` but should be considered as invalid, whereas `Next[<non-existing element> == 0xffffffffffffffffffffffffffffffffffffffff` returns `false` which means that this address is not in the list.

**Comment by akolotov** [comment]
> Currently, we have a "validators" boolStorage. Shall we discard it in favor of this new validator list? Or should we keep and maintain both?

No, we can consider the mapping for validators to be the address storage and keep the reward addresses there. If a validator exists, it means that it's reward address is not `address(0)`

**Comment by fernandomg** [comment]
@akolotov,

as expected, `while` is cheaper for short lists. But I think that we should go with the double-linked-list as it's more scalable.

@patitonar, thoughts?

```shell
·------------------------------------------------------------------------------------------------|-----------------------------------·
|                                              Gas                                               ·  Block limit: 17592186044415 gas  │
·································································|·······························|····································
|  Methods                                                       ·          1 gwei/gas           ·          110.85 usd/eth           │
·······························|·································|·········|·········|···········|··················|·················
|  Contract                    ·  Method                         ·  Min    ·  Max    ·  Avg      ·  # calls         ·  usd (avg)     │
·······························|·································|·········|·········|···········|··················|·················
|  BridgeValidators            ·  initialize                     ·      -  ·      -  ·  2826248  ·              50  ·          0.31  │
·······························|·································|·········|·········|···········|··················|·················
|  BridgeValidators            ·  removeValidatorFromLinkedList  ·  22548  ·  71188  ·    44181  ·              50  ·          0.00  │
·······························|·································|·········|·········|···········|··················|·················
|  BridgeValidatorsLinkedList  ·  initialize                     ·      -  ·      -  ·  3918462  ·              50  ·          0.43  │
·······························|·································|·········|·········|···········|··················|·················
|  BridgeValidatorsLinkedList  ·  removeValidatorFromDLinkedList ·  28581  ·  28613  ·    28611  ·              50  ·          0.00  │
·······························|·································|·········|·········|···········|··················|·················
|  Deployments                                                   ·                               ·  % of limit      ·                │
·································································|·········|·········|···········|··················|·················
|  BridgeValidators                                              ·      -  ·      -  ·  1570315  ·             0 %  ·          0.17  │
·································································|·········|·········|···········|··················|·················
|  BridgeValidatorsLinkedList                                    ·      -  ·      -  ·  1720470  ·             0 %  ·          0.19  │
·----------------------------------------------------------------|---------|---------|-----------|------------------|----------------·
```

**Comment by fernandomg** [comment]
@akolotov,

The `single-linked-list` starts to be more expensive from the validator nr 13.

![singlelinkedlist-vs-doublelinkedlist](https://user-images.githubusercontent.com/3315606/50298799-c7617280-045e-11e9-827c-42323126b320.png)

**Comment by akolotov** [comment]
just curious. Could you send gist with both implementations?

**Comment by fernandomg** [comment]
[Single Linked List](https://gist.github.com/fernandomg/9a03b039eb1e91c47f424bc5f468b81d)

[Double Linked List](https://gist.github.com/fernandomg/186776a0b775d30b5039973373406676)

**Comment by akolotov** [comment]
Available in [the release 2.3.0](https://github.com/poanetwork/poa-bridge-contracts/releases/tag/2.3.0)
